File: CPP/ByTier/A/0790_A_Bear_and_Different_Names.cpp
// Problem: CF 790 A - Bear and Different Names
// https://codeforces.com/contest/790/problem/A

/* 
 * Purpose: This program solves the "Bear and Different Names" problem.
 *          Given constraints on consecutive groups of soldiers' names,
 *          it generates a valid sequence of names such that:
 *          - Each group of k consecutive soldiers has all unique names (YES)
 *          - Or at least one name is repeated (NO)
 *          Names follow the format: First letter uppercase, rest lowercase.
 *          
 * Algorithm:
 * - Start with a base set of names using a simple pattern (like "Aa", "Ab", ..., "Az", "Ba", ...),
 *   ensuring no name is too long and all follow the naming rule.
 * - For each group described as "NO", we adjust the last soldier's name in that group
 *   to be identical to the first soldier's name, violating the uniqueness requirement.
 * 
 * Time Complexity: O(n*k) where n is number of soldiers and k is group size.
 *                  We iterate through n-k+1 groups, and for each, we may copy a name.
 * Space Complexity: O(n) for storing the names array.
 */

#include <stddef.h>
#include <iostream>
#include <string>

using namespace std;

signed main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  
  int n, k;
  cin >> n >> k;
  
  // Initialize names with a simple pattern (e.g., Aa, Ab, Ac, ..., Az, Ba, ..., Zz)
  string ans[n];
  for (int i = 0; i < n; i++) {
    ans[i] = "00";                      // Initialize with two characters
    ans[i][0] = i / 26 + 'A';           // First letter: A-Z using integer division
    ans[i][1] = i % 26 + 'a';           // Second letter: a-z using modulo
  }
  
  string s;
  // Process each group description from input
  for (int i = 0; i <= n - k; i++) {
    cin >> s;
    if (s == "NO") {
      // If group is not effective, make the last soldier's name same as first soldier's name
      ans[i + k - 1] = ans[i];
    }
  }
  
  // Output all generated names
  for (string i : ans)
    cout << i << ' ';
    
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/0790/0790_A_Bear_and_Different_Names.py
# Problem: CF 790 A - Bear and Different Names
# https://codeforces.com/contest/790/problem/A

import string


def list_to_string(_a):
    return " ".join(map(str, _a))


def solve():
    # Precompute possible names: single uppercase letters and two-letter combinations
    # starting with uppercase and followed by lowercase
    possible_names = list(string.ascii_uppercase)
    possible_names += list(
        map(
            lambda x: x[0] + x[1],
            zip(list(string.ascii_uppercase), list(string.ascii_lowercase)),
        )
    )
    n, k = map(int, input().split())
    a = list(input().split())
    result = possible_names[:n]  # Start with basic names
    # Iterate through each consecutive group of k soldiers
    for i in range(n - k + 1):
        if a[i] == "NO":  # If the group is not effective
            # Make the last soldier of the group same as the first soldier
            # This ensures that the group will have duplicate names (hence "NO")
            result[i + k - 1] = result[i]
    return list_to_string(result)


if __name__ == "__main__":
    print(solve())

### Complexity Analysis

# - **Time Complexity:** O(n * k), where n is the number of soldiers and k is the group size. We iterate through the groups and in worst case we may need to adjust the names of soldiers in each group.

# - **Space Complexity:** O(n), as we store the names of soldiers and precompute a list of possible names which has a fixed size dependent on the alphabet.

# https://github.com/VaHiX/CodeForces/